### PR TITLE
risk acceptance: enhance!

### PR DESCRIPTION
**Pull request by valentijnscholten**
Risk Acceptance enhancements as per #3441.

Adds fields:
- make decision on risk treatment (accept, reduce, avoid, transfer)
- add details field to explain decision (was compensating controls)
- add recommendation field ~~(optional)~~ for security team to advise
- add details field to explain recommendations
- handle expiration of risk acceptances
- optionally findings are reactivated
- optionally the sla period is restarted
- the expiration cron job posts comments to jira on expiration (optional) and as dojo alert.
- the expiration cron job posts a X days heads up warning before expiration (optional) (to jira and as dojo alert)
- management command to trigger the expiration handler manually: `./manage.py risk_acceptance_handle_expiration`
- post a comment to jira when a finding is accepted or unaccepted
- full status sync with JIRA was and is still not in place because it would require more information about the workflow and statuses and will only work for very simple workflows. feel free to add more status syncing.
- allow editing of risk acceptances
- allow manual ad hoc expiration and/or reinstating of risk acceptances
image

### CODE REVIEW

**Comment by madchap** [comment]
fyi @yilmi

File: dojo/db_migrations/0068_risk_acceptance.py
Comment (madchap) [inline dojo/db_migrations/0068_risk_acceptance.py:135]: Rephrase to "Date at which notice about the risk acceptance expiration was sent"?

File: dojo/db_migrations/0068_risk_acceptance.py
Comment (madchap) [inline dojo/db_migrations/0068_risk_acceptance.py:130]: Missing end parentheses in description.

File: dojo/forms.py
Comment (madchap) [inline dojo/forms.py:575]: Should we accept MS office document as well?

File: dojo/db_migrations/0068_risk_acceptance.py
Comment (madchap) [inline dojo/db_migrations/0068_risk_acceptance.py:68]: I tend to write "DefectDojo" and not "defect dojo". I guess that's something we should clarify with other maintainers too.

**Review by madchap — CHANGES_REQUESTED** [review]
I was wondering if the Product Overview page should eventually output a summary of risk acceptance made across the various engagements?

Also, regarding the "risk actions", I usually use the following. The "fix" is missing, and "reduce" and "compensate" should probably be merged into "mitigate"?

image

The proof upload could possibly give headaches, just like for images and such. We still have ways to go for docker-compose, and still nothing for k8s.

Looks good otherwise :-)

**Comment by madchap** [comment]
If uploading no proof at all, it still shows

image

Clicking on None causes a 500 error.

File: dojo/db_migrations/0068_risk_acceptance.py
Comment (madchap) [inline dojo/db_migrations/0068_risk_acceptance.py:68]: I would not put that `an entity` can accept the risk. Must be a person, for accountability purposes. Should mention the person accepting the risk must have authority to do so.

**Review by madchap — APPROVED** [review]
lgtm, tested locally, vast improvement :-)

**Comment by madchap** [comment]
A couple conflicts @valentijnscholten ?

File: dojo/filters.py
Comment (alles-klar) [inline dojo/filters.py:1387]: This code lines are the same as above?

Comment (valentijnscholten) [reply dojo/filters.py]: good find, fixed

File: dojo/models.py
Comment (alles-klar) [inline dojo/models.py:1391]: can be deleted if it is not used anymore?

Comment (valentijnscholten) [reply dojo/models.py]: done

**Comment by alles-klar** [comment]
Great improvement @valentijnscholten !

Todos before merging from my point of view:
- please remove debug output which is commented out for cleaner code
- resolve conflicts and fix tests
- can a product have full and simple risk acceptance enabled in parallel or can you choose only one? 
- at the moment it is possible to add a simple risk acceptance also when the flag in the product is disabled
- "add recommendation field (optional) for security team to advise" is not optional in the ui
- did you tested what happens which old "full risk acceptances" when upgrading to this version?
- I am missing some tests, especially for expiration and the sla reset

Some ideas for improvements - not sure if you want to implement them in this pr:
- Add Decision Option "No Decision" for situations where you want to write down the acceptance but didn't got feedback from your management. Could be an inactive acceptance until you get the decision.
- As @madchap suggested a summary of risks in the Product Overview or maybe also generally summary over all products

**Comment by valentijnscholten** [comment]
> * please remove debug output which is commented out for cleaner code
I will leave it for now cause there's no easy way to find it. I will do cleanup PR later with lots of comments removed from historical prs.

> * resolve conflicts and fix tests
done

> * can a product have full and simple risk acceptance enabled in parallel or can you choose only one?
can both be enabled, up to the user to decide

> * at the moment it is possible to add a simple risk acceptance also when the flag in the product is disabled
how can I reproduce this, for me it works as it should

> * "add recommendation field (optional) for security team to advise" is not optional in the ui
I think non-optional is actually better.

> * did you tested what happens which old "full risk acceptances" when upgrading to this version?
yes and so did madchap. did you test also?

> * I am missing some tests, especially for expiration and the sla reset
they are in `test_risk_acceptance.py`, or what are you missing specifically?

**Comment by alles-klar** [comment]
> > * at the moment it is possible to add a simple risk acceptance also when the flag in the product is disabled
> 
> how can I reproduce this, for me it works as it should
> 
I have a product form an older defectdojo version and simple risk acceptances are disabled (by default). In the findings view I can use bulk edit to accept findings (or are bulk risk acceptance another category? If yes we should simplify it...)

> > * did you tested what happens which old "full risk acceptances" when upgrading to this version?
> 
> yes and so did madchap. did you test also?
no

> > * I am missing some tests, especially for expiration and the sla reset
> 
> they are in `test_risk_acceptance.py`, or what are you missing specifically?
I can only find test_bulk_risk_acceptance_api.py in unittests?

**Comment by valentijnscholten** [comment]
> I have a product form an older defectdojo version and simple risk acceptances are disabled (by default). In the findings view I can use bulk edit to accept findings (or are bulk risk acceptance another category? If yes we should simplify it...)

Hmm might be a bug indeed, wasn't aware it was about the bulk acceptance. Will check.

> > > * I am missing some tests, especially for expiration and the sla reset
> > 
> > 
> > they are in `test_risk_acceptance.py`, or what are you missing specifically?
> 
> I can only find test_bulk_risk_acceptance_api.py in unittests?

hmm looks like with all the merging and rebasing it resetted some of the new files. will re-add.

**Comment by valentijnscholten** [comment]
Thanks for the thorough review @alles-klar. Re-added the missing files (tests) and moved the permission check to the risk acceptance helper so permissions are always checked to see if simple risk acceptance is enabled.

**Comment by valentijnscholten** [comment]
hmmm the bulk edit is not straightforward. there can be products with simple risk acceptance enabled mixed with disabled.

**Comment by valentijnscholten** [comment]
have now updated the code to skip findings that cannot be "simple risk accepted" and show a warning afterwards:
image

File: dojo/engagement/views.py
Comment (alles-klar) [inline dojo/engagement/views.py:1080]: why is this commented out?

**Review by alles-klar — APPROVED** [review]

File: dojo/engagement/views.py
Comment (valentijnscholten) [reply dojo/engagement/views.py:1080]: old code that I kept in comments because I wasn't sure if I would need it later or not. it's no longer needed. I will clean it up in the cleanup PR later. for now I want to merge this PR after so many hours :-)

**Pull request merged**
